### PR TITLE
Access control power up

### PR DESCRIPTION
**Pull request by mmadariaga**
Add the ability to parse inheritance on child access control nodes
Reviewed API access control directives (fixed or added)
Added a new kind of tests (access control tests) as part of API testing
Updated fixtures and API tests so that they pass new access rules

### CODE REVIEW

File: web/rest/client/tests/DataAccessControl/Provider/BillableCallTest.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    function it_has_write_access_control()
+    {
+        $accessControl = $this
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/BillableCallTest.php]: Deny access

File: web/rest/client/tests/DataAccessControl/Provider/CallForwardSettingTest.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            [
+                [
+                    'or' => [
+                        [
+                            'user',
+                            'in',
+                            'UserRepository([["company","eq","user.getCompany().getId()"]])'
+                        ],
+                        [
+                            'user',
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/CallForwardSettingTest.php]: combine conditions and remove nulls

File: web/rest/client/tests/DataAccessControl/Provider/CallForwardSettingTest.php [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'or' => [
+                        [
+                            'user',
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/CallForwardSettingTest.php]: combine with residentialDevice and retailAccount

File: web/rest/client/tests/DataAccessControl/Provider/ConditionalRoutesConditionsRelRouteLockTest.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'routeLock',
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/ConditionalRoutesConditionsRelRouteLockTest.php]: + conditionalRouteCondition validation

File: web/rest/client/tests/DataAccessControl/Provider/ConditionalRoutesConditionsRelScheduleTest.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::READ_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'condition',
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/ConditionalRoutesConditionsRelScheduleTest.php]: ScheduleRepository will be faster

File: web/rest/client/tests/DataAccessControl/Provider/DdiTest.php [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                        [
+                            'queue',
+                            'isNull',
+                            null
+                        ]
+                    ]
+                ]
+            ]
+        );
+    }
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/DdiTest.php]: + RetailAccount, ResidentialDevice, ConditionalRoute and ExternalCallFilter

File: web/rest/client/tests/DataAccessControl/Provider/FaxTest.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'company',
+                    'eq',
+                    'user.getCompany().getId()'
+                ]
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/FaxTest.php]: + OutgoingDdi

File: web/rest/client/tests/DataAccessControl/Provider/FeaturesRelCompanyTest.php [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'company',
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/FeaturesRelCompanyTest.php]: Deny access

File: web/rest/client/tests/DataAccessControl/Provider/FriendTest.php [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                            null
+                        ]
+                    ]
+                ],
+                [
+                    'or' => [
+                        [
+                            'outgoingDdi',
+                            'in',
+                            'OutgoingDdiRuleRepository([["company","eq","user.getCompany().getId()"]])'
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/FriendTest.php]: DdiRepository

File: web/rest/client/tests/DataAccessControl/Provider/FriendTest.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                        ],
+                        [
+                            'outgoingDdi',
+                            'isNull',
+                            null
+                        ]
+                    ]
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/FriendTest.php]: + transformationRuleSet

File: web/rest/client/tests/DataAccessControl/Provider/InvoiceTest.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'company',
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/InvoiceTest.php]: Deny access

File: web/rest/client/tests/DataAccessControl/Provider/IvrExcludedExtensionTest.php [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'ivr',
+                    'in',
+                    'IvrRepository([["company","IN",["CompanyRepository([[\"id\",\"eq\",\"user.getCompany().getId()\"]])"]]])'
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/IvrExcludedExtensionTest.php]: \+ Extension

File: web/rest/client/tests/DataAccessControl/Provider/MatchListPatternTest.php [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'matchList',
+                    'in',
+                    'MatchListRepository({"or":[["brand","eq","user.getCompany().getBrand().getId()"],["company","eq","user.getCompany().getId()"]]})'
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/MatchListPatternTest.php]: Not the brand onces

File: web/rest/client/tests/DataAccessControl/Provider/MusicOnHoldTest.php [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+                DataAccessControlParser::READ_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
+                    'or' => [
+                        [
+                            'brand',
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/MusicOnHoldTest.php]: unnecessary

File: web/rest/client/tests/DataAccessControl/Provider/OutgoingDdiRulesPatternTest.php [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+                    'and' => [
+                        [
+                            'outgoingDdiRule',
+                            'in',
+                            'OutgoingDdiRuleRepository([["company","eq","user.getCompany().getId()"]])'
+                        ],
+                        [
+                            'matchList',
+                            'in',
+                            'MatchListRepository({"or":[["brand","eq","user.getCompany().getBrand().getId()"],["company","eq","user.getCompany().getId()"]]})'
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/OutgoingDdiRulesPatternTest.php]: nullable

File: web/rest/client/tests/DataAccessControl/Provider/QueueMemberTest.php [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+                    'or' => [
+                        [
+                            'user',
+                            'in',
+                            'UserRepository([["company","eq","user.getCompany().getId()"]])'
+                        ],
+                        [
+                            'user',
+                            'isNull',
+                            null
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/QueueMemberTest.php]: Do not allow null

File: web/rest/client/tests/DataAccessControl/Provider/QueueMemberTest.php [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+                    'or' => [
+                        [
+                            'queue',
+                            'in',
+                            'QueueRepository([["company","IN",["CompanyRepository([[\"id\",\"eq\",\"user.getCompany().getId()\"]])"]]])'
+                        ],
+                        [
+                            'queue',
+                            'isNull',
+                            null
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/QueueMemberTest.php]: Do not allow null

File: web/rest/client/tests/DataAccessControl/Provider/RatingProfileTest.php [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    function it_has_write_access_control()
+    {
+        $accessControl = $this
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/RatingProfileTest.php]: Deny access

File: web/rest/client/tests/DataAccessControl/Provider/ResidentialDeviceTest.php [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+                            'CompanyRepository([["id","eq","user.getCompany().getId()"]])'
+                        ]
+                    ]
+                ],
+                [
+                    'or' => [
+                        [
+                            'outgoingDdi',
+                            'in',
+                            'OutgoingDdiRuleRepository([["company","eq","user.getCompany().getId()"]])'
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/ResidentialDeviceTest.php]: DDI

File: web/rest/client/tests/DataAccessControl/Provider/ResidentialDeviceTest.php [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+                        ],
+                        [
+                            'outgoingDdi',
+                            'isNull',
+                            null
+                        ]
+                    ]
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/ResidentialDeviceTest.php]: \+ TransformationRuleSet

File: web/rest/client/tests/DataAccessControl/Provider/RetailAccountTest.php [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+                            'CompanyRepository([["id","eq","user.getCompany().getId()"]])'
+                        ]
+                    ]
+                ],
+                [
+                    'or' => [
+                        [
+                            'outgoingDdi',
+                            'in',
+                            'OutgoingDdiRuleRepository([["company","eq","user.getCompany().getId()"]])'
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/RetailAccountTest.php]: DDI

File: web/rest/client/tests/DataAccessControl/Provider/RetailAccountTest.php [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+                        ],
+                        [
+                            'outgoingDdi',
+                            'isNull',
+                            null
+                        ]
+                    ]
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/RetailAccountTest.php]: \+ TransformationRuleSet

File: web/rest/client/tests/DataAccessControl/Provider/RoutingTagTest.php [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+        $accessControl = $this
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
+                [
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/RoutingTagTest.php]: Deny access

File: web/rest/client/tests/DataAccessControl/Provider/RoutingTagTest.php [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    function it_has_read_access_control()
+    {
+        $accessControl = $this
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::READ_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/RoutingTagTest.php]: Deny access

File: web/rest/client/tests/DataAccessControl/Provider/TerminalModelTest.php [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $accessControl = $this
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/TerminalModelTest.php]: Deny Access

File: web/rest/client/tests/DataAccessControl/Provider/TimezoneTest.php [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $accessControl = $this
+            ->dataAccessControlParser
+            ->get(
+                DataAccessControlParser::WRITE_ACCESS_CONTROL_ATTRIBUTE
+            );
+
+        $this->assertEquals(
+            $accessControl,
+            [
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/TimezoneTest.php]: Deny access

File: web/rest/client/tests/DataAccessControl/Provider/UserTest.php [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+                            null
+                        ]
+                    ]
+                ],
+                [
+                    'or' => [
+                        [
+                            'outgoingDdi',
+                            'in',
+                            'OutgoingDdiRuleRepository([["company","eq","user.getCompany().getId()"]])'
```
Comment (mmadariaga) [inline web/rest/client/tests/DataAccessControl/Provider/UserTest.php]: Ddi

**Comment by ironArt3mis** [comment]
Rebase completed.

**Pull request merged**